### PR TITLE
[MeshModel] Add `WASM Filter` as a core component.

### LINKED ISSUE #7903: [MeshModel] Support WASM Filters as a Core Model
#### Current Behavior
Simply filters currently follows MeshOps v1 style support.

#### Desired Behavior
Support WASM Filters as a core MeshModel model.

#### Implementation
1. Create "Meshery" Core Model - (store in Kubernetes model?)
1. Create "Meshery Filter" Component
1. Create Relationship Definition - type: hierarchical, sub-type: wallet, parent: Istio Envoy Filter.
1. Create Policy Definition - 
1. Create Context-Aware - Update the Istio Envoy Filter Patch with the config from the Meshery Filter component.

#### Acceptance Tests
1. Meshery Filter component is ever-present and relatable to Istio Envoy Filters.

---
#### Contributor [Guides](https://docs.meshery.io/project/contributing) and [Handbook](https://layer5.io/community/handbook)
- 🛠 [Meshery Build & Release Strategy](https://docs.meshery.io/project/contributing/build-and-release)
- 📚 [Instructions for contributing to documentation](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#documentation-contribution-flow)
   - Meshery documentation [site](https://docs.meshery.io/) and [source](https://github.com/meshery/meshery/tree/master/docs)
- 🎨 Wireframes and designs for Meshery UI in [Figma](https://www.figma.com/file/SMP3zxOjZztdOLtgN4dS2W/Meshery-UI)
- 🙋🏾🙋🏼 Questions: [Discussion Forum](https://discuss.layer5.io) and [Community Slack](http://slack.layer5.io)

### PR DESCRIPTION
**Pull request by MUzairS15**
**Notes for Reviewers**

This PR fixes #7903
1. Adds `WASM Filter` as a `Core` component under the `Orchestration & Management` Category.

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

File: server/meshmodel/components/core/WASMFilter.json [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
...7.98 16.17zM127 99.37a67.22 67.22 0 0 0 7.91-28.94l-29.13 16.68Z\"\/><path class=\"cls-1\" d=\"m103.64 83.69 28.12-16.08-28.12-16.16v32.24z\"\/><path class=\"cls-2\" d=\"M99.61 44.5V12.52L71.76 28.49 99.61 44.5zm0 39.05V51.28L71.7 67.44l27.91 16.11zm-32.13 51.47h.01-.01z\"\/><path class=\"cls-1\" d=\"M35.54 51.22v32.51l28.12-16.28-28.12-16.23z\"\/><path d=\"M65.47 0a67.2 67.2 0 0 0-29.64 7.83l29.64 17Z\" class=\"cls-2\"\/><path class=\"cls-1\" d=\"M35.54 12.3v32.32l28.14-16.14L35.54 12.3z\"\/><path d=\"M31.51 10.34A67.89 67.89 0 0 0 10.1 31.89l21.41 12.36Z\" class=\"cls-2\"\/><path d=\"M99.43 8a67.23 67.23 0 0 0-29.94-8v25.15ZM0 69.87a67.27 67.27 0 0 0 8.07 29.76l21.69-12.56Zm8.07-34.5A67.16 67.16 0 0 0 0 65l29.79-17.09Z\" class=\"cls-1\"\/><path d=\"M35.78 127.13A67.13 67.13 0 0 0 65.47 135v-24.85Z\" class=\"cls-2\"\/><path d=\"M124.92 32a67.9 67.9 0 0 0-21.28-21.52V44.3Zm-21.28 92.54A68 68 0 0 0 125 102.86l-21.36-12.34Z\" class=\"cls-1\"\/><path d=\"M135 64.81a67.06 67.06 0 0 0-8-29.35l-21.51 12.42Z\" class=\"cls-2\"\/><path d=\"M69.49 135a67.12 67.12 0 0 0 29.63-7.83L69.49 110Z\" class=\"cls-1\"\/><path class=\"cls-2\" d=\"M31.51 83.44V51.56L3.83 67.43l27.68 16.01z\"\/><\/svg>",
```
Comment (leecalcote) [inline server/meshmodel/components/core/WASMFilter.json:15]: Please use the webassembly logo, not Meshery.

**Review by leecalcote — CHANGES_REQUESTED** [review]
We are spreadsheet driven in terms of ComponentDefinitions. Please define the component in Meshery Integrations.

**Comment by MUzairS15** [comment]
@leecalcote yes, reason for not adding in Integration spreadsheet is, this PR is not complete and I don’t know adding this new component may start breaking things hence didn’t added.

I will add checklist for what’s remaining

**Comment by MUzairS15** [comment]
- [ ] Relationship Definition
- [ ] Policy for defining context aware behaviour for WASM Filter component

**Comment by leecalcote** [comment]
> @leecalcote yes, reason for not adding in Integration spreadsheet is, this PR is not complete and I don’t know adding this new component may start breaking things hence didn’t added.
> 
> I will add checklist for what’s remaining

Make sense. For the new ComponentDefinition, we do need to define in the spreadsheet. You’re all clear to do so immediately. 👍 call me if you have questions about the sheet. This is an area that you need to become very familiar with as you go to fill-in for @Revolyssup, who has very little time for this any longer. 

On a related note, the component generation framework in MeshKit, should be setup as a factory system in which support for adding helm repos other than Artifact Hub is possible. There are a many more than helm repos to be added soon.

**Review by Abhishek-kumar09 — COMMENTED** [review]
Questions:
1. Should all wasm filters contains non-empty config?
2. Filters is an array, do each individual can also contain config and entrypoints?
3. Is there any working prototype that this configuration would work?

**Comment by MUzairS15** [comment]
> Questions:
> 
> 1. Should all wasm filter contains non-empty config?
> 2. Filters is an array, do each individual can also contain config and entrypoints?
> 3. Is there any working prototype that this configuration would work?

#1 IMO it should
#2 no it’s a mistake it should be a string 
#3 Only the envoy filter will be deployed and the confit will populated appropriately from WASMFilter component as part of ctx aware policy so any valid Envoy filter will work

**Comment by MUzairS15** [comment]
@leecalcote @Abhishek-kumar09 Can we move forward here?

File: server/meshmodel/components/core/WASMFilter.json [hunk 2]
```diff
@@ -0,0 +1,33 @@
+{
+    "kind": "WASMFilter",
+    "apiVersion": "core.meshmodel.dev/v1alpha1",
+    "displayName": "WASM Filter",
+    "format": "JSON",
+    "isAnnotation": true,
+    "metadata": {
+        "logo-URL": "",
```
Comment (leecalcote) [inline server/meshmodel/components/core/WASMFilter.json:8]: Change to logoURL to maintain our camelcasing design principle. If we have done this in other components, hopefully we don’t have logic expecting a hyphen.

File: server/meshmodel/components/core/WASMFilter.json [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
...7.98 16.17zM127 99.37a67.22 67.22 0 0 0 7.91-28.94l-29.13 16.68Z\"\/><path class=\"cls-1\" d=\"m103.64 83.69 28.12-16.08-28.12-16.16v32.24z\"\/><path class=\"cls-2\" d=\"M99.61 44.5V12.52L71.76 28.49 99.61 44.5zm0 39.05V51.28L71.7 67.44l27.91 16.11zm-32.13 51.47h.01-.01z\"\/><path class=\"cls-1\" d=\"M35.54 51.22v32.51l28.12-16.28-28.12-16.23z\"\/><path d=\"M65.47 0a67.2 67.2 0 0 0-29.64 7.83l29.64 17Z\" class=\"cls-2\"\/><path class=\"cls-1\" d=\"M35.54 12.3v32.32l28.14-16.14L35.54 12.3z\"\/><path d=\"M31.51 10.34A67.89 67.89 0 0 0 10.1 31.89l21.41 12.36Z\" class=\"cls-2\"\/><path d=\"M99.43 8a67.23 67.23 0 0 0-29.94-8v25.15ZM0 69.87a67.27 67.27 0 0 0 8.07 29.76l21.69-12.56Zm8.07-34.5A67.16 67.16 0 0 0 0 65l29.79-17.09Z\" class=\"cls-1\"\/><path d=\"M35.78 127.13A67.13 67.13 0 0 0 65.47 135v-24.85Z\" class=\"cls-2\"\/><path d=\"M124.92 32a67.9 67.9 0 0 0-21.28-21.52V44.3Zm-21.28 92.54A68 68 0 0 0 125 102.86l-21.36-12.34Z\" class=\"cls-1\"\/><path d=\"M135 64.81a67.06 67.06 0 0 0-8-29.35l-21.51 12.42Z\" class=\"cls-2\"\/><path d=\"M69.49 135a67.12 67.12 0 0 0 29.63-7.83L69.49 110Z\" class=\"cls-1\"\/><path class=\"cls-2\" d=\"M31.51 83.44V51.56L3.83 67.43l27.68 16.01z\"\/><\/svg>",
```
Comment (leecalcote) [inline server/meshmodel/components/core/WASMFilter.json:27]: @Revolyssup do you know why were defining model information twice? We see this duplication in a number of other components.

File: server/meshmodel/components/core/WASMFilter.json [hunk 4]
```diff
@@ -0,0 +1,33 @@
+{
+    "kind": "WASMFilter",
+    "apiVersion": "core.meshmodel.dev/v1alpha1",
+    "displayName": "WASM Filter",
+    "format": "JSON",
+    "isAnnotation": true,
```
Comment (leecalcote) [inline server/meshmodel/components/core/WASMFilter.json:6]: Sliding this into metadata makes sense.

File: server/meshmodel/relationships/hierarchical_wallet.json [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        "name": "core",
+        "version": "v1.0.0",
+        "displayName": "Core",
+        "category": {
+            "name": "Orchestration \u0026 Management",
+            "metadata": null
+        },
+        "metadata": {}
+    },
+    "subType": "Parent",
```
Comment (leecalcote) [inline server/meshmodel/relationships/hierarchical_wallet.json:17]:
```suggestion
    "subType": "Inventory",    
```

File: server/meshmodel/components/core/WASMFilter.json (hunk 2 above)
Comment (leecalcote) [reply server/meshmodel/components/core/WASMFilter.json:8]: @Abhishek-kumar09 do you know?

Comment (leecalcote) [inline server/meshmodel/components/core/WASMFilter.json:8]:
```suggestion
```

File: server/meshmodel/components/core/WASMFilter.json [hunk 6]
```diff
@@ -0,0 +1,33 @@
+{
+    "kind": "WASMFilter",
+    "apiVersion": "core.meshmodel.dev/v1alpha1",
```
Comment (leecalcote) [inline server/meshmodel/components/core/WASMFilter.json:3]:
```suggestion
    "apiVersion": "core.meshery.io/v1alpha1",
```

File: server/meshmodel/components/core/WASMFilter.json [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    "displayName": "WASM Filter",
+    "format": "JSON",
+    "isAnnotation": true,
+    "metadata": {
+        "logo-URL": "",
+        "model": "core",
+        "modelDisplayName": "Core",
```
Comment (leecalcote) [inline server/meshmodel/components/core/WASMFilter.json:9]:
```suggestion
        "model": "kubernetes",
        "modelDisplayName": "Kubernetes",
```

File: server/meshmodel/components/core/WASMFilter.json [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+        }
+    },
+    "schema": "{\"$id\":\"http:\/\/meshery.layer5.io\/definition\/Workload\/WASMFilter\",\"$schema\":\"http:\/\/json-schema.org\/draft-07\/schema\",\"properties\":{\"wasm-filter\":{\"type\":\"string\"},\"config\":{\"type\":\"string\"}},\"required\":[\"wasm-filter\",\"config\"],\"title\":\"WASM Filter\",\"type\":\"object\"}"
```
Comment (leecalcote) [inline server/meshmodel/components/core/WASMFilter.json:30]:
```suggestion
    "schema": "{\"$id\":\"https:\/\/meshery.layer5.io\/system\/models\/definition\/Workload\/WASMFilter\",\"$schema\":\"http:\/\/json-schema.org\/draft-07\/schema\",\"properties\":{\"wasm-filter\":{\"type\":\"string\"},\"config\":{\"type\":\"string\"}},\"required\":[\"wasm-filter\",\"config\"],\"title\":\"WASM Filter\",\"type\":\"object\"}"
```

File: server/meshmodel/relationships/hierarchical_wallet.json [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+            "to": [
+                {   "kind": "WASMFilter",
+                    "model": "Core"
+                }
+            ]
+        },
+        "deny": {
+            "from": [],
+            "to": []
+        }
```
Comment (leecalcote) [inline server/meshmodel/relationships/hierarchical_wallet.json:36]:
```suggestion
    "subType": "Inventory",    
    "referenceURI":
    "selectors": {
        "allow": {
          "from": [
              {
                  "kind": "EnvoyFilter",
                  "model": "Istio Base",
                  "patch": {
                    "patchStrategy": "replace",
                    "$mutatedRef": "#configPatches.patch.value",
                    "description": "Receive the WASM filter configuration."
                  }
              }
          ],
          "to": [
              {   "kind": "WASMFilter",
                  "model": "Core",
                  "patch": {
                    "patchStrategy": "replace",
                    "$mutatorRef": "#config",
                    "description": "WASM filter configuration to be applied to Envoy Filter."
                  }
                }
          ],
        },
        "deny": {
            "from": [],
            "to": []
        }
```

File: server/meshmodel/relationships/hierarchical_wallet.json [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+        "name": "core",
+        "version": "v1.0.0",
+        "displayName": "Core",
+        "category": {
+            "name": "Orchestration \u0026 Management",
+            "metadata": null
+        },
+        "metadata": {}
+    },
+    "subType": "Inventory",
```
Comment (leecalcote) [inline server/meshmodel/relationships/hierarchical_wallet.json:17]: A new subtype of "Inventory" to be defined in for MeshModel Relationships.

File: server/meshmodel/relationships/hierarchical_wallet.json [hunk 11]
```diff
@@ -0,0 +1,38 @@
+{
+    "apiVersion": "core.meshery.io/v1alpha1",
+    "kind": "Hierarchical",
+    "metadata": {
+        "description": "A relationship that defines whether a component can be a parent of other components. It is read like: `from` can be a parent of `to`, where from and to are any components that match the from and to selectors respectively"
+    },
+    "model": {
+        "name": "core",
+        "version": "v1.0.0",
+        "displayName": "Core",
```
Comment (leecalcote) [inline server/meshmodel/relationships/hierarchical_wallet.json:10]:
```suggestion
        "description": "A hierarchical inventory relationship in which a Core WASM Filter configuration is applied to an Istio Envoy Filter."
    },
    "model": {
        "name": "kubernetes",
        "version": "*",
        "displayName": "Kubernetes",
```

**Review by leecalcote — APPROVED** [review]

**Comment by leecalcote** [comment]
> @leecalcote @Abhishek-kumar09 Can we move forward here?

LGTM. Thanks for the in-depth reviews. We have three new schema changes coming out of this: 

1. Relationships of subType Inventory
2. Components Metadata of isAnnotation
3. Relationships To/From Patches

Two open questions, @Revolyssup, @Abhishek-kumar09:

1. Why is "logo-URL" not camelcased? I recognize that we don't use the field currently, but it needs to follow spec.
2. Why is "model" repeated twice in the same Component Definition - once under metadata and once under model attribute?

**Comment by leecalcote** [comment]
One follow up here: the overall conversion of WASM Filters into a MeshModel component, inclusive of binary reference.

**Pull request merged**